types/logger/logger.go

```diff
@@ -114,6 +121,10 @@ func RateLimitedFn(logf Logf, f time.Duration, burst int, maxCache int) Logf {
 			}
 		}
 
+		if len(contexts) > 0 {
+			format += " (rate-limit-context:" + strings.Join(contexts, ",") + ")"
```

Comment:
this seems like a lot of allocs per Logf.

Reply:
All of this is really expensive. This line is only a single alloc. But there are also two allocs in logfExtra (a string concatenation and a slice alloc), and two allocs from the strings.Replacer call.

I'll add a benchmark suite in a final commit. Results (per log entry):

```
BenchmarkRateLimiter-8          	 6015066	       194 ns/op	       0 B/op	       0 allocs/op
BenchmarkNoRateLimit-8          	10655953	       107 ns/op	      96 B/op	       2 allocs/op
BenchmarkRateLimiterContext-8   	 2654502	       451 ns/op	     176 B/op	       5 allocs/op
BenchmarkStdLogger-8            	 4168108	       286 ns/op	      80 B/op	       3 allocs/op
BenchmarkPostProcess-8          	 2268715	       529 ns/op	     136 B/op	       8 allocs/op
```